### PR TITLE
8342540: InterfaceCalls micro-benchmark gives misleading results

### PR DESCRIPTION
**Pull request by theRealAph**
`InterfaceCalls.java` makes highly predictable memory accesses, which leads to a gross time underestimate of the case where a megamorphic access is unpredictable.

Here's one example, with and without randomization. The unpredictable megamorphic call takes more than 4* as long as the benchmark.

```
Benchmark                            (randomized)  Mode  Cnt   Score   Error  Units
InterfaceCalls.test2ndInt3Types             false  avgt    4   5.013 ± 0.081  ns/op
InterfaceCalls.test2ndInt3Types              true  avgt    4  23.421 ± 0.102  ns/op
``` 

This patch adds the "randomized" parameter, which allows the measurement of predictable and unpredictable megamorphic calls.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342540](https://bugs.openjdk.org/browse/JDK-8342540): InterfaceCalls micro-benchmark gives misleading results (**Enhancement** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21581/head:pull/21581` \
`$ git checkout pull/21581`

Update a local copy of the PR: \
`$ git checkout pull/21581` \
`$ git pull https://git.openjdk.org/jdk.git pull/21581/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21581`

View PR using the GUI difftool: \
`$ git pr show -t 21581`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21581.diff">https://git.openjdk.org/jdk/pull/21581.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21581#issuecomment-2422304581)

### CODE REVIEW

File: test/micro/org/openjdk/bench/vm/compiler/InterfaceCalls.java [hunk 1]
```diff
@@ -45,6 +46,10 @@
 @Fork(value = 3)
 public class InterfaceCalls {
 
+    // Whether to step iteratively through the list of interfaces, or
+    // to select one in an unpredictable way.
+    @Param({"false", "true"}) private boolean randomized;
```
Comment (shipilev) [inline test/micro/org/openjdk/bench/vm/compiler/InterfaceCalls.java:51]:
```suggestion
    @Param({"false", "true"})
    private boolean randomized;
```

**Comment by theRealAph** [comment]
> Is there even a point to do non-randomized test then?

That's a very interesting question, and one that has been occupying me since I discovered this problem a few days ago.

I have no idea how often well-predicted megamorphic calls occur. I can speculate that the "typical" megamorphic case lies somewhere between these extremes, but that is all. It may well be that normal behaviour is chaotic, but I strongly suspect that the cases are unlikely to be equally probable, as they are here. So, it is possible that an utterly unpredictable access pattern is just as unrealistic as a perfectly predictable one.

**Comment by shipilev** [comment]
> I have no idea how often well-predicted megamorphic calls occur. I can speculate that the "typical" megamorphic case lies somewhere between these extremes, but that is all. It may well be that normal behaviour is chaotic, but I strongly suspect that the cases are unlikely to be equally probable, as they are here. So, it is possible that an utterly unpredictable access pattern is just as unrealistic as a perfectly predictable one.

Yeah, all right. We can have a test that brackets the real-world performance between "best" and "worst" cases.

If you want to make "best" and "worst" only differ in the actual payload, maybe we should compute both scrambled and non-scrambled index, feed both to `Blackhole`, and only then select the index based on `randomized`. This way we would always compute both indexes.

Also stats question: Do you know if scrambling actually produces the full period in requested `range`, and whether the frequency for individual cases is roughly the same? We kinda assume this xor-step is well distributed and has good enthropy in lower bits, but is it in practice?

**Comment by theRealAph** [comment]
> > I have no idea how often well-predicted megamorphic calls occur. I can speculate that the "typical" megamorphic case lies somewhere between these extremes, but that is all. It may well be that normal behaviour is chaotic, but I strongly suspect that the cases are unlikely to be equally probable, as they are here. So, it is possible that an utterly unpredictable access pattern is just as unrealistic as a perfectly predictable one.
> 
> Yeah, all right. We can have a test that brackets the real-world performance between "best" and "worst" cases.
> 
> If you want to make "best" and "worst" only differ in the actual payload, maybe we should compute both scrambled and non-scrambled indexes, feed both to `Blackhole`, and only then select the index based on `randomized`. This way we would always compute both indexes.

Sure, I can try that, but I doubt it'll affect much. The great virtue of the xorshift generator is to be so lightweight, just a few clocks, that it barely affects anything.

> Also stats question: Do you know if scrambling actually produces the full period in requested `range`, and whether the frequency for individual cases is roughly the same? We kinda assume this xor-step is well distributed and has good enthropy in lower bits, but is it in practice?

I don't think we assume it: the analysis is in Marsaglia's _Xorshift RNGs_ paper. The full period (except 0) is proven there. 

The generator I've used here is a compromise: it has to be just good enough to throw off a branch predictor, while still being of very low overhead. There are several elaborations of Xorshift with better statistical properties (see "xoroshiro*") but they are less successful at just getting out of the way.

**Review by vnkozlov — APPROVED** [review]
Good.

**Pull request closed without merging**